### ISSUE TITLE
TRPO not supports GaussianLSTMPolicy

### ISSUE THREAD

**Issue by ilinojkovic** [rlworkgroup/garage#741]
Just by changing the line with `GaussianMLPPolicy` to `GaussianLSTMPolicy` in `trpo_swimmer.py` example I get the error:
```
TypeError: Second-order gradient for while loops not supported.
```
Same goes for `GaussianLSTMPolicyWithModel` and `GaussianGRUPolicy`.

Is this the expected behavior and should TRPO support recurrent policies?

**Comment by ryanjulian** [comment]
@ahtsan please investigate

**Comment by ryanjulian** [comment]
@ilinojkovic thanks for reporting! we will take a look

**Comment by ilinojkovic** [comment]
No problem. Hopefully it won't be a hustle to resolve.

**Comment by ahtsan** [comment]
@ilinojkovic Good catch! Try replacing `algo = ...` with
```
        algo = TRPO(
            env_spec=env.spec,
            policy=policy,
            baseline=baseline,
            max_path_length=500,
            discount=0.99,
            max_kl_step=0.01,
            optimizer=ConjugateGradientOptimizer,
            optimizer_args=dict(
                hvp_approach=FiniteDifferenceHvp(base_eps=1e-5)))
```

(Ref: https://github.com/rlworkgroup/garage/blob/master/examples/tf/trpo_cartpole_recurrent.py)

`ConjugateGradientOptimizer` and `FiniteDifferenceHvp` should be used for recurrent policies.

**Comment by ilinojkovic** [comment]
Hi @ahtsan, thanks for the response. I tried your suggestion, but now I get a different error:
```
AttributeError: Can't pickle local object 'FiniteDifferenceHvp.update_opt.<locals>.<lambda>'
```
This breaks when snapshot is being saved.

**Comment by ahtsan** [comment]
@ilinojkovic We are working on the issue. As a quick fix, comment `snapshotter.save_itr_params(epoch, params)` in `local_tf_runner.py`.

**Comment by ilinojkovic** [comment]
Do you have an ETA on the fix? I'd kinda need the snapshots, so to figure if it's worth to hack my way out or wait for official change.

**Comment by ahtsan** [comment]
@ilinojkovic Should be done soon, blocked on the CI. You can look at the changes in https://github.com/rlworkgroup/garage/pull/745.

**Comment by ilinojkovic** [comment]
Thanks!

**Comment by ryanjulian** [comment]
@ahtsan oer @ilinojkovic -- please add a unit test which would have caught this bug. i think you can just run one step of TRPO using GaussianLSTMPolicy, right?

**Comment by ryanjulian** [comment]
Fixed by #745. please (re)open a new issue if this isn't solved by the most up-to-date code.

**Comment by ahtsan** [comment]
@ilinojkovic Please kindly check if the issue is resolved.
